### ISSUE TITLE
issue about _parent_value for xgboost trees

### ISSUE THREAD

**Issue by SauceCat** [TeamHG-Memex/eli5#155]
I have checked through the codes in xgboost.py and became quite confused about this part of codes:
```python
def _parent_value(children):
    """ Value of the parent node: a weighted sum of child values.
    """
    covers = np.array([child['cover'] for child in children])
    covers /= np.sum(covers)
    leafs = np.array([child['leaf'] for child in children])
    return np.mean(leafs * covers)
```
I think `leafs * covers` already represents the weighted sum of child values, because `covers` have already been normalized by `covers /= np.sum(covers)`. Thus, I wonder why should this function return `np.mean(leafs * covers)` instead of `leafs * covers` ?

**Comment by lopuhin** [comment]
Thanks for the question, @SauceCat !
We need a single value for the parent node, but ``leafs * covers`` is an element-wise multiplication, so we need to reduce it to a scalar:
```
In [1]: leafs
Out[1]: array([ 0.0222222, -0.191398 ])

In [2]: covers
Out[2]: array([ 0.05319149,  0.94680851])

In [3]: leafs * covers
Out[3]: array([ 0.00118203, -0.18121726])

In [4]: np.mean(leafs * covers)
Out[4]: -0.090017611702127673
```
``np.mean`` is used to get the value (instead of e.g. ``np.sum``) because we want the expected value of the parent, where "probability" is given by ``cover``.

**Comment by SauceCat** [comment]
Hi @lopuhin , thanks for replying! 
Yes, indeed, what I mean is to return np.sum(leafs * cover) instead of np.mean (leafs * cover).
I am confused because I have tried to modify the codes of XGBoost, and make `get_dump` function also output the `base_weight` on each tree node(including leaf nodes and parent nodes). You can check my post here: https://github.com/dmlc/xgboost/issues/1986.   
I found that `leaf_value = base_weight * learning_rate`, thus value of each parent node can also be calculated as `base_weight * learning_rate`. Compared my calculation with eli5's, I found results from eli5 package was always 0.5 * myresult. Thus I thought maybe the average is not necessary here.   

> cover in parent node = cover in left leaf + cover in right leaf
> value of parent node = cover in left leaf / cover in parent node * value of left leaf + cover in right leaf / cover in parent node * value of right leaf  

Why we need average here?

**Comment by lopuhin** [comment]
@SauceCat I think you are right, to get the expected value the sum is needed, not the mean, because sum of covers is already 1. Thanks a lot for pointing that out!
Changing mean to sum leaves feature contribution sum the same (which was already correct), but changes individual feature contributions. Now sure yet how to properly test it.

**Comment by lopuhin** [comment]
By the way, visualisation here https://github.com/dmlc/xgboost/issues/1986#issue-202723776 is very nice and informative 👍

**Comment by kmike** [comment]
@lopuhin what do you think about stealing colors @SauceCat used for the visualization?
